src/app/core/testing/utils/mock-storage-generator.utils.ts

```diff
@@ -0,0 +1,381 @@
+    const disks: StorageDashboardDisk[] = [];
+
+    for (let i = 0; i < repeats; i++) {
+      const vdev: VDev = this.generateVdev(layout, width) as VDev;
+      vdev.children.forEach((child: TopologyDisk) => {
+        const disk = this.generateDisk(diskSize, (disks.length + allDisks.length));
+        child.disk = disk.name;
+        child.stats = { size: disk.size } as TopologyItemStats;
+        disks.push(disk);
+      });
+
+      vdev.stats.size = this.calculateVdevCapacity(vdev);
+      vdevs.push(vdev);
+    }
+
+    return {
+      topologyItems: vdevs,
+      disks,
+    };
+  }
+
+  private generateNoRedundancyTopology(repeats: number,
+    diskSize: number,
+    width: number,
+    allDisks: StorageDashboardDisk[] = this.disks): MockTopology {
+    const vdevs: TopologyItem[] = [];
+    const disks: StorageDashboardDisk[] = [];
+
+    for (let i = 0; i < repeats; i++) {
+      const vdev: TopologyDisk = this.generateVdev(TopologyItemType.Disk, width) as TopologyDisk;
+      const disk = this.generateDisk(diskSize, (disks.length + allDisks.length));
+      vdev.disk = disk.name;
+      vdevs.push(vdev);
+      disks.push(disk);
+    }
+
+    return {
+      topologyItems: vdevs,
+      disks,
+    };
+  }
+
+  private generateMixedDiskCapacityTopology(layout: TopologyItemType,
+    repeats: number,
+    diskSize: number,
+    width: number,
+    allDisks: StorageDashboardDisk[] = this.disks): MockTopology {
+    const vdevs: TopologyItem[] = [];
+    const disks: StorageDashboardDisk[] = [];
+
+    for (let i = 0; i < repeats; i++) {
+      const vdev: VDev = this.generateVdev(layout, width) as VDev;
+      vdev.children.forEach((child: TopologyDisk, index) => {
+        const childDiskSize = index === 0 ? diskSize + 1 : diskSize;
+        const disk = this.generateDisk(childDiskSize, (disks.length + allDisks.length));
+        child.disk = disk.name;
+        child.stats = { size: disk.size } as TopologyItemStats;
+        disks.push(disk);
+      });
+
+      vdev.stats.size = this.calculateVdevCapacity(vdev);
+      vdevs.push(vdev);
+    }
+
+    return {
+      topologyItems: vdevs,
+      disks,
+    };
+  }
+
+  // Generate VDEV
+  private generateVdev(layout: TopologyItemType, width = 1): TopologyDisk | VDev {
+    const minWidth = this.getMinLayoutWidth(layout);
+    if (width < minWidth) {
+      width = minWidth;
+    }
+
+    if (layout === TopologyItemType.Disk || layout === TopologyItemType.Stripe) {
+      return this.generateTopologyDisk();
+    }
+
+    switch (layout) {
+      case TopologyItemType.Raidz3:
+      case TopologyItemType.Raidz2:
+      case TopologyItemType.Raidz1:
+      case TopologyItemType.Raidz:
+      case TopologyItemType.Mirror: {
+        const vdevDisks: TopologyDisk[] = [];
+        for (let i = 0; i < width; i++) {
+          const topologyDisk: unknown = this.generateTopologyDisk();
+          vdevDisks.push(topologyDisk as TopologyDisk);
+        }
+
+        const vdev: unknown = {
```

Comment:
`const vdev = {} as Vdev.`
`unknown` will swallow errors, like if you mistype one of the fields.

Same in other places in this file.